### DISCUSSION TITLE
Alternatives for ElasticSearch

### DISCUSSION THREAD

**Discussion by thomas-advantitge** [vendurehq/vendure#734] [General]
We already talked about this once, discussing [MeiliSearch](https://www.meilisearch.com/) as a possible alternative for ElasticSearch for example. 

Currently, ElasticSearch requires more than 3 times the resources of the Vendure server + worker application. 
Next to that, it's not meant to be run in serverless environments which would be an advantage for running preview/non-production environments. We currently need to provision a VM with 2Gi memory for every ElasticSearch instance, after disabling default replication which would need even more resources.

As I didn't find an existing thread yet, starting a discussion here to hear if there are concrete plans for the future supporting other powerful search mechanisms than Elastic.

**Comment by michaelbromley** [comment]
MeiliSearch currently has the limitation that query-time sorting is not supported. So you can't have a search results page and then choose to sort by price (ASC, DESC), name, etc. To me that's a show-stopper, but they have an open issue for it: https://github.com/meilisearch/MeiliSearch/issues/730. In a recent interaction on Twitter they said they plan it for Q3 20201.

I'm also looking into [TypeSense](https://github.com/typesense/typesense) which _does_ support sorting, and seems to be otherwise comparable to MeiliSearch.

**Comment by michaelbromley** [comment]
I've been doing some research into this today. Here's a feature comparison table I put together. Although I am really drawn to both Meilisearch and TypeSense, they are still lacking some features which I would consider essential in an e-commerce search. TypeSense is very promising but lacks aggregations - so we'd need some other way to generate pricing information from results, or drop that feature. 


Feature | Elastic | MeiliSearch | TypeSense
-- | -- | -- | --
Faceted search | ✅ | ✅ [docs](https://docs.meilisearch.com/reference/features/faceted_search.html#filters-or-facets) | ✅ [docs](https://typesense.org/docs/0.19.0/guide/building-a-search-application.html#faceting)
Run-time sort | ✅ | ❌ | ✅
Aggregations | ✅ | ❌ | ❌
Synonyms | ✅ [docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.12/analysis-synonym-tokenfilter.html) | ✅ [docs](https://docs.meilisearch.com/reference/features/synonyms.html#synonyms) | ✅ [docs](https://typesense.org/docs/0.19.0/api/synonyms.html)
Curation (promoted results) | ✅[docs](https://www.elastic.co/guide/en/app-search/7.9/curations.html) | ❌ | ✅ [docs](https://typesense.org/docs/0.19.0/api/curation.html)

Here's another really great comparison between TypeSense, Algolia, Elastic & Meilisearch: https://typesense.org/typesense-vs-algolia-vs-elasticsearch-vs-meilisearch/

**Reply by mariusa** [reply]
Is there an issue request at TypeSense for Aggregations?

**Reply by michaelbromley** [reply]
I reached out via Twitter and it seems it might be possible to do what we need using the faceting functionality

https://twitter.com/typesense/status/1382734835370786817?s=19

**Comment by stefanvanherwijnen** [comment]
I am having some trouble trying to figure out how ElasticSearch is used/integrated. If I understand correctly, it essentially has its own database which is optimized for indexing/searching? If so, does the ElasticSearchPlugin perform the work of keeping everything in sync? (e.g. when I add a product it gets added to both the database and to ElasticSearch)

**Reply by michaelbromley** [reply]
Yes, Elasticsearch is a specialized document database which is optimized for searching large datasets quickly and provides lots of powerful search-related features. The Vendure ElasticSearchPlugin does indeed take care of keeping the search index in sync with your product data. It is a drop-in replacement for the DefaultSearchPlugin - which also uses a separate index but the difference being it is stored as a table in your primary DB and it's search capabilities are much more limited than with Elasticsearch.

**Reply by stefanvanherwijnen** [reply]
Thanks,  I couldn't wrap my head around on how everything was kept in sync, but having to take care of that yourself makes sense. I'll have a closer look at the plugin for some implementation details which might come in handy in one of my projects which requires full text search (as the current SQL LIKE solution definitely isn't user friendly).

I also stumbled upon your YT video about this subject which cleared up some things : https://youtu.be/Zv1BVT9FgJo
Looking forward for the AdvancedSearchPlugin 👍
